### PR TITLE
Dj query api

### PR DESCRIPTION
**Pull request by CircArgs**
### Summary

This pr aims to add the ability to execute via the api a "DJ SQL" that allows otherwise illegal references to dj nodes. For example

Example:
```SQL
    SELECT Avg(n_comments),
        age
    FROM   (SELECT basic.num_comments            AS n_comments,
                basic.dimension.users.country country,
                basic.dimension.users.age     AS age
            FROM   metrics
            GROUP  BY basic.dimension.users.country, basic.dimension.users.age)
    GROUP  BY age
```

Get's translated from the route to an intermediate query that would be a legal dj node query and ulitmately built into this thing

```sql
SELECT  Avg(n_comments) AS col0,
        age
 FROM (SELECT  basic_DOT_num_comments.cnt AS n_comments,
        basic_DOT_dimension_DOT_users.country AS country,
        basic_DOT_dimension_DOT_users.age AS age
 FROM (SELECT  COUNT(1) AS cnt,
        basic.comments.id AS basic_DOT_source_DOT_comments_DOT_id,
        basic.comments.user_id AS basic_DOT_source_DOT_comments_DOT_user_id,
        basic.comments.timestamp AS basic_DOT_source_DOT_comments_DOT_timestamp,
        basic.comments.text AS basic_DOT_source_DOT_comments_DOT_text
 FROM basic.comments

) AS basic_DOT_num_comments
LEFT JOIN basic.comments
        ON basic_DOT_num_comments.basic_DOT_source_DOT_comments_DOT_id = basic.comments.id AND basic_DOT_num_comments.basic_DOT_source_DOT_comments_DOT_user_id = basic.comments.user_id AND basic_DOT_num_comments.basic_DOT_source_DOT_comments_DOT_timestamp = basic.comments.timestamp AND basic_DOT_num_comments.basic_DOT_source_DOT_comments_DOT_text = basic.comments.text
LEFT JOIN (SELECT  basic.comments.id,
        basic.comments.full_name,
        basic.comments.age,
        basic.comments.country,
        basic.comments.gender,
        basic.comments.preferred_language,
        basic.comments.secret_number
 FROM basic.comments

) AS basic_DOT_dimension_DOT_users
        ON basic.comments.user_id = basic_DOT_dimension_DOT_users.id AND basic.comments.user_id = basic_DOT_dimension_DOT_users.id
 GROUP BY  basic_DOT_dimension_DOT_users.country, basic_DOT_dimension_DOT_users.age)

 GROUP BY  age
```
which successfully runs and returns a small table. 

The hope is for this to also be an escape hatch for users to test different things on the fly.

### Test Plan

- [ ] PR has an associated issue: #
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### CODE REVIEW

**Review by agorajek — COMMENTED** [review]
Whoa, this looks cool @CircArgs, but I am wondering how did the content of dj_query.py got past our linters ;)

File: dj/construction/dj_query.py
Comment (samredai) [inline dj/construction/dj_query.py:18]: I love where this is going! Some early feedback is I think we should split the validation in this function out to a separate function and then call that in a `validate_dj_query` function in `dj/api/helpers.py` and even a `"/sql/validate/"` endpoint.

File: dj/construction/dj_query.py
Comment (agorajek) [inline dj/construction/dj_query.py:44]: `s/'metrics.'/'metrics'`

File: dj/construction/dj_query.py
Comment (agorajek) [inline dj/construction/dj_query.py:56]: `s/for a Metric select/in a Metric query`

File: tests/cli/urls_test.py
Comment (agorajek) [inline tests/cli/urls_test.py:33]: Pls disregard the location of this comment, its just easy to see here. These new paths are confusing next to the existing `/queries/` path. How about ?
```
/queries/data/: Return data for a DJ Query.
/queries/sql/: Return SQL for a DJ Query.
```
Or did I miss the fact that these are special queries and you made a separate path on purpose?

**Review by agorajek — COMMENTED** [review]
I also like where this is going :) Left few comments in line.

File: tests/cli/urls_test.py
Comment (samredai) [reply tests/cli/urls_test.py:33]: Agreed the current paths are confusing but I also feel that `"/queries/sql/"` will be confusing to new-comers who don't fully understand giving SQL to get different SQL. Since we'll need an endpoint for validating a query without running it, maybe we can offer that as a way to get the built sql? Then we wouldn't need to specify `data` in the endpoint.
```
POST /query/: Return data for a DJ Query
POST /query/validate/: Validate a DJ Query
```
The second one would also return the built SQL. Thoughts?

Comment (samredai) [reply tests/cli/urls_test.py:33]: Also I noticed the `GET "/queries/{query_id}/"` endpoint which should still work as expected. For consistency with my suggestions above, we should rename that to `GET "/query/{query_id}/"` (if we go with my suggestions, that is 
😄 ).

Comment (agorajek) [reply tests/cli/urls_test.py:33]: I like your suggestion @samredai . Especially since I am not a fan of the REST paths starting with a plural term. I am not familiar with the REST standards so I won't push but imo this:
```
.../thing/{thing_id}
.../thing/list
```
is more intuitive than:
```
.../things/{thing_id}
.../things/
```

Comment (CircArgs) [reply tests/cli/urls_test.py:33]: @samredai so is this correct?

1. `/query`: the user sends a str query to get data
2. `/query/validate`: the user sends a str query to see if it is valid and gets back the query we would run
3. `/query/{query_id}`: I don't actually know what this does haha

and so I should make the `/sql` routes I have right now conform to 1, 2?

Comment (shangyian) [reply tests/cli/urls_test.py:33]: Isn't `/query/{query_id}` more for tracking the results of a specific query? It would only make sense after someone has actually kicked off a query run. One potential flow:
* `/query` -> user sends query str, DJ returns a reference to a `query_id` that identifies the query run
* `/query/validate` -> users sends query str to see if it's valid
* `/query/{query_id}` -> returns the status of the query run
* `/query/{query_id}/data` -> returns the query data

Comment (CircArgs) [reply tests/cli/urls_test.py:33]: @shangyian that looks very sensible to me!

Comment (samredai) [reply tests/cli/urls_test.py:33]: @CircArgs yep that's right. And @shangyian good catch, it makes sense to me to split out the data retrieval part to a `/query/{query_id}/data/` endpoint. Right now it just shows up in the status response when the status is "COMPLETED" but that's not ideal. I like the explicit follow-up call to get the actual data.

File: dj/api/query.py
Comment (samredai) [inline dj/api/query.py:2]: s/Metric/DJ SQL

File: dj/api/query.py
Comment (samredai) [inline dj/api/query.py:26]: This should be `router.post` since it requires a body

File: dj/construction/dj_query.py
Comment (samredai) [inline dj/construction/dj_query.py:50]: Not sure what this comment means, could use some clarification/elaboration

File: dj/construction/dj_query.py
Comment (samredai) [inline dj/construction/dj_query.py:89]: nit: Since the loop just calls the `_hoist_metric_source_tables`, these comments should probably just go in that function's doc-string

**Review by samredai — APPROVED** [review]
LGTM! Just a few comments but approving to unblock you. I think this PR could use a squash as well before merging.

**Pull request merged**
